### PR TITLE
IC-proxy support for gpexpand

### PR DESCRIPTION
**Pull request by Aegeaner**
Adjust that `gp_interconnect_proxy_addresses` GUC during gpexpand, making the expanding operation atomic in IC proxy mode. 

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gpexpand [hunk 1]
```diff
@@ -1140,6 +1167,71 @@ class gpexpand:
...
+            self.conn = dbconn.connect(self.dburl, utility=True, encoding='UTF8', allowSystemTableMods=True)
+        sql = "show gp_interconnect_proxy_addresses;"
+        icProxyAddressesStr = dbconn.queryRow(self.conn, sql)[0]
+        self._parseIcProxyAddressStr(icProxyAddressesStr)
+        outputfile = 'gpexpand_ic_proxy_backup.bak'
+        logger.info("Backing up IC proxy addresses...")
+        with open(outputfile, 'w') as outf:
+            outf.write(icProxyAddressesStr)
+
+    def _parseIcProxyAddressStr(self, icProxyAddressesStr):
```
Comment (kainwen) [inline gpMgmt/bin/gpexpand:1218]: `_parseIcProxyAddressStr` is started with a singe `_`, but `__encodeIcProxyAddress` with two.

And I think we do not need to introduce these two functions. The logic here is quite simple.

Comment (kainwen) [inline gpMgmt/bin/gpexpand:1218]: `_parseIcProxyAddressStr` is started with a singe `_`, but `__encodeIcProxyAddress` with two.

And I think we do not need to introduce these two functions. The logic here is quite simple.

File: gpMgmt/bin/gppylib/operations/unix.py [hunk 2]
```diff
@@ -43,7 +43,11 @@ def execute(self):
                       pickled=True)
         cmd.run(validateAfter=True)
         # TODO! If exception is raised remotely, there's no stdout, thereby causing a pickling error.
-        return pickle.loads(cmd.get_results().stdout)
+        try:
```
Comment (kainwen) [inline gpMgmt/bin/gppylib/operations/unix.py:46]: Why this change?

Comment (kainwen) [inline gpMgmt/bin/gppylib/operations/unix.py:46]: Why this change?

File: gpMgmt/bin/gpexpand [hunk 3]
```diff
@@ -1140,6 +1167,71 @@ class gpexpand:
         cursor = dbconn.query(self.conn, sql)
         return dict(cursor.fetchall())
 
+    def check_interconnect_type(self):
+        self.get_interconnect_proxy_addresses()
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply gpMgmt/bin/gpexpand:1171]: But the `interconnect_type` may still be configured as udpifc or other, even in proxy mode.

File: gpMgmt/bin/gpexpand [hunk 4]
```diff
@@ -1140,6 +1167,71 @@ class gpexpand:
...
+                                'gp_interconnect_proxy_addresses',
+                                value="'" + icProxyAddrStr + "'",
+                                ctxt=REMOTE,
+                                remoteHost=seg.hostname)
+            pool.addCommand(cmd)
+        pool.join()
+        pool.check_results()
+        pool.haltWork()
+        pool.joinWorkers()
+        os.system('gpstop -u')
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply gpMgmt/bin/gpexpand:1205]: `gpstop -u` will sighup the whole cluster. The worker pool is managing each cmd's execution, outside the CMD.

File: gpMgmt/bin/gpexpand (hunk 3 above)
[the remark this answers is not in the document]
Comment (kainwen) [reply gpMgmt/bin/gpexpand:1171]: > But the `interconnect_type` may still be configured as udpifc or other, even in proxy mode.

Correct. I mean check the GUC `gp_interconnect_proxy_addresses`.

File: gpMgmt/bin/gpexpand (hunk 1 above)
Comment (Aegeaner) [reply gpMgmt/bin/gpexpand:1218]: `_parseIcProxyAddressStr` is used two times, we shouldn't repeat ourselves even if logic is simple.

File: gpMgmt/bin/gpexpand [hunk 5]
```diff
@@ -950,6 +960,8 @@ class gpexpand:
             self.tempDir = createTempDirectoryName(self.options.coordinator_data_directory, "gpexpand")
         self.queue = None
         self.segTemplate = None
+        self.icProxyAddresses = []
```
[the remark this answers is not in the document]
Comment (Aegeaner) [reply gpMgmt/bin/gpexpand:963]: we generate a recommended config file from the segment configuration, it's structured.
while the proxy address we read from config file is just a string.
need to transform between the structured object and the written string, the `icProxyAddresses` is a temp intermediate variable

File: gpMgmt/bin/gpexpand (hunk 4 above)
[the remark this answers is not in the document]
Comment (kainwen) [reply gpMgmt/bin/gpexpand:1205]: > `gpstop -u` will sighup the whole cluster. The worker pool is managing each cmd's execution, outside the CMD.

I mean, we might write like this:

```
cmd = Command(xxx, xxx, "gpstop -u")
cmd.run(validateAfter=True)
...
```

**Comment by kainwen** [comment]
Close this:

* long time no update
* need more design

Will create some internal stories:

* what current behavior and what the correct behavior when setting port is used by other applications
* based on the above conclusion, design the UI of specifying port of new added segments and push forward this.
